bitcoin/examples/sighash.rs

```diff
@@ -0,0 +1,199 @@
+use bitcoin::{consensus, ecdsa, sighash, PublicKey, Script, ScriptBuf, Transaction};
+use hex_lit::hex;
+
+//These are real blockchain transactions examples of computing sighash for:
+// - P2WPKH
+// - P2MS 2of3
+// - P2SH 2of2 multisig
+// - P2WSH 2of2 multisig
+
+//run with: cargo run --example sighash
+
+//TODO add P2TR examples, ideally for both key-path and script-path spending
+
+/// Computes segwit sighash for a transaction input that spends a p2wksh output with "witness_v0_keyhash" scriptPubKey.type
+///
+/// # Arguments
+///
+/// * `raw_tx` - spending tx hex
+/// * `inp_idx` - spending tx input index
+/// * `value` - ref tx output value in sats
+fn compute_sighash_p2wpkh(raw_tx: &[u8], inp_idx: usize, value: u64) {
+    let tx: Transaction = consensus::encode::deserialize(raw_tx).unwrap();
+    let inp = &tx.input[inp_idx];
+    let witness = &inp.witness;
+    println!("Witness: {:?}", witness);
+
+    // BIP-141: The witness must consist of exactly 2 items (≤ 520 bytes each). The first one a
+    // signature, and the second one a public key.
+    assert_eq!(witness.len(), 2);
+    let sig_bytes = witness.nth(0).unwrap();
+    let pk_bytes = witness.nth(1).unwrap();
+
+    let sig = ecdsa::Signature::from_slice(sig_bytes).expect("failed to parse sig");
+
+    //BIP-143: "The item 5 : For P2WPKH witness program, the scriptCode is 0x1976a914{20-byte-pubkey-hash}88ac"
+    //this is nothing but a standard P2PKH script OP_DUP OP_HASH160 <pubKeyHash> OP_EQUALVERIFY OP_CHECKSIG:
+    let pk = PublicKey::from_slice(pk_bytes).expect("failed to parse pubkey");
+    let wpkh = pk.wpubkey_hash().expect("compressed key");
+    println!("Script pubkey hash: {:x}", wpkh);
+    let spk = ScriptBuf::new_v0_p2wpkh(&wpkh);
+    let script_code = spk.p2wpkh_script_code().expect("Spk is known to be p2wpkh");
+
+    let mut cache = sighash::SighashCache::new(&tx);
+    let sighash = cache
+        .segwit_signature_hash(inp_idx, &script_code, value, sig.hash_ty)
+        .expect("failed to compute sighash");
+    println!("Segwit p2wpkh sighash: {:x}", sighash);
+    let msg = secp256k1::Message::from(sighash);
+    println!("msg is {:x}", msg);
```

Comment:
Just for uniformity with the rest :)
```suggestion
    println!("Message: {:x}", msg);
```